server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -23,30 +23,43 @@
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.common.util.concurrent.ThreadContext;
 
+import java.util.ArrayList;
+import java.util.List;
+
 /**
  * A logger that logs deprecation notices. Logger should be initialized with a parent logger which name will be used
  * for deprecation logger. For instance <code>new DeprecationLogger("org.elasticsearch.test.SomeClass")</code> will
  * result in a deprecation logger with name <code>org.elasticsearch.deprecation.test.SomeClass</code>. This allows to use
  * <code>deprecation</code> logger defined in log4j2.properties.
- *
+ * <p>
  * Deprecation logs are written to deprecation log file - defined in log4j2.properties, as well as warnings added to a response header.
  * All deprecation usages are throttled basing on a key. Key is a string provided in an argument and can be prefixed with
  * <code>X-Opaque-Id</code>. This allows to throttle deprecations per client usage.
  * <code>deprecationLogger.deprecate("key","message {}", "param");</code>
- *
- * @see ThrottlingAndHeaderWarningLogger for throttling and header warnings implementation details
+ * <p>
+ * Additional log handlers can be registered via {@link #addHandler(DeprecatedLogHandler)}. Subject to rate limiting,
+ * each additional handler will also be called with a deprecated message and its key.
  */
 public class DeprecationLogger {
-    private final ThrottlingAndHeaderWarningLogger deprecationLogger;
+    /** Log handlers are are not tied to this specific <code>DeprecationLogger</code> instance */
+    private static final List<DeprecatedLogHandler> additionalHandlers = new ArrayList<>();
```

Comment:
will adding/remove handlers be done in another thread? Or do we expect this to be set-up before any logging is allowed? (pre logconfig is loaded)
I wonder if this should be thread safe.
Also why not keeping them both together in the same collection?. Looks like the iteration in `withDeprecation` is the same (at least at the moment)